learning_observer/learning_observer/lms_integration.py

```diff
@@ -0,0 +1,287 @@
+    - target_url: The URL endpoint to be called, with optional formatting using kwargs.
+    - base_url: An optional base URL for the service. If provided, it will be prefixed 
+      to target_url.
+    - kwargs: Additional keyword arguments to format the target_url or control behavior 
+      (e.g., retry).
+
+    Returns:
+    - A JSON response from the requested service, with key translation if necessary.
+
+    Raises:
+    - aiohttp.web.HTTPUnauthorized: If the request lacks necessary authorization.
+    - aiohttp.ClientResponseError: If the request fails, with special handling for 401 errors on Canvas.
+    """
+    # Retrieve the incoming request and active user
+    request = runtime.get_request()
+    user = await learning_observer.auth.get_active_user(request)
+    
+    # Extract 'retry' flag from kwargs (defaults to False)
+    retry = kwargs.pop('retry', False)
+    
+    # Retrieve session and determine the appropriate headers based on the service
+    session = await aiohttp_session.get_session(request)
+    headers = {
+        'google': request[constants.AUTH_HEADERS],
+        'canvas': session.get(constants.CANVAS_AUTH_HEADERS)
+    }
+
+    # Ensure Google requests are authenticated
+    if lms_name == 'google' and constants.AUTH_HEADERS not in request:
+        raise aiohttp.web.HTTPUnauthorized(text="Please log in")
+    
+    # Construct the full URL using the base URL if provided, otherwise use the target URL directly
+    if base_url:
+        url = base_url + target_url.format(**kwargs)
+    else:
+        url = target_url.format(**kwargs)
+
+    # Generate a unique cache key based on the service, user, and request URL
+    cache_key = f"raw_{lms_name}/" + learning_observer.auth.encode_id('session', user[constants.USER_ID]) + '/' + learning_observer.util.url_pathname(url)
+    
+    # Check cache and return cached response if available
+    if settings.feature_flag(f"use_{lms_name}_ajax") is not None:
+        value = await cache[cache_key]
+        if value is not None:
+            # Translate keys if the service is Google, otherwise return raw JSON
+            if lms_name == 'google':
+                return learning_observer.util.translate_json_keys(
+                    json.loads(value),
+                    learning_observer.google.GOOGLE_TO_SNAKE
+                )
+            else: 
+                return json.loads(value)
+
+    # Make the actual AJAX call to the service
+    async with aiohttp.ClientSession(loop=request.app.loop) as client:
+        try:
+            async with client.get(url, headers=headers[lms_name]) as resp:
+                response = await resp.json()
+
+                # Log the AJAX request and response
+                learning_observer.log_event.log_ajax(target_url, response, request)
+                # Cache the response if the feature flag is enabled
+                if settings.feature_flag(f"use_{lms_name}_ajax") is not None:
```

Comment:
Create a variable for `f"user_{lms_name}_ajax"` since its used in multiple places.

Reply:
Done